sdks/java/core/src/main/java/org/apache/beam/sdk/coders/DelegateCoder.java

```diff
@@ -101,8 +104,28 @@ public Object structuralValue(T value) throws Exception {
   }
 
   @Override
+  public boolean equals(Object o) {
+    if (o == null || this.getClass() != o.getClass()) {
```

Comment:
Cool article.

I suggest that `DelegateCoder` is not intended for subclassing, so we could make it final. Then I think what Pei has is the same as what Luke suggests here. I think that we should actually move more towards the "different types cannot be equal" perspective throughout the SDK, unless those two types are both defined to be just implementations of a common interface, and then the interface should own the definition of equality.